pkg/controllers/disruption/disruptionbudgetmapping.go

```diff
@@ -0,0 +1,184 @@
+	if reasonBudget, ok := d.mapping[nodePool][disruptionReason]; ok {
+		return reasonBudget
+	}
+	return d.mapping[nodePool][v1beta1.DisruptionReasonDefault]
+}
+
+// BuildDisruptionBudgetMapping will build the disruption budget mapping. The disruption budget maps for each disruption reason the number of allowed disruptions for each node pool.
+// We calculate allowed disruptions by taking the max disruptions allowed by disruption reason and subtracting the number of nodes that are already being deleted by that disruption reason.
+//
+//nolint:gocyclo
+func BuildDisruptionBudgetMapping(ctx context.Context, cluster *state.Cluster, clk clock.Clock, kubeClient client.Client, recorder events.Recorder) (*DisruptionBudgetMapping, error) {
+	disruptionBudgetMapping := NewDisruptionBudgetMapping(map[string]map[v1beta1.DisruptionReason]int{})
+	numNodes := map[string]int{}   // map[nodepool] -> node count in nodepool
+	disrupting := map[string]int{} // map[nodepool] -> nodes undergoing disruption
+	for _, node := range cluster.Nodes() {
+		// We only consider nodes that we own and are initialized towards the total.
+		// If a node is launched/registered, but not initialized, pods aren't scheduled
+		// to the node, and these are treated as unhealthy until they're cleaned up.
+		// This prevents odd roundup cases with percentages where replacement nodes that
+		// aren't initialized could be counted towards the total, resulting in more disruptions
+		// to active nodes than desired, where Karpenter should wait for these nodes to be
+		// healthy before continuing.
+		if !node.Managed() || !node.Initialized() {
+			continue
+		}
+
+		nodePool := node.Labels()[v1beta1.NodePoolLabelKey]
+		numNodes[nodePool]++
+
+		// If the node satisfies one of the following, we subtract it from the allowed disruptions.
+		// 1. Has a NotReady conditiion
+		// 2. Is marked as disrupting
+		if cond := nodeutils.GetCondition(node.Node, v1.NodeReady); cond.Status != v1.ConditionTrue || node.MarkedForDeletion() {
+			disrupting[nodePool]++
+		}
+	}
+	nodePoolList := &v1beta1.NodePoolList{}
+	if err := kubeClient.List(ctx, nodePoolList); err != nil {
+		return nil, fmt.Errorf("listing node pools, %w", err)
+	}
+	for _, nodePool := range nodePoolList.Items {
+		minDisruptionsByReason, err := nodePool.GetAllowedDisruptionsByReason(ctx, clk, numNodes[nodePool.Name])
+		if err != nil {
+			return nil, fmt.Errorf("getting allowed disruptions by reason, %w", err)
+		}
+		allowedDisruptionsTotal := 0
+		for reason, minDisruptions := range minDisruptionsByReason {
+
+			// Subtract the allowed number of disruptions from the number of already disrupting nodes.
+			// Floor the value since the number of disrupting nodes can exceed the number of allowed disruptions.
+			// Allowing this value to be negative breaks assumptions in the code used to calculate how many nodes can be disrupted.
+			allowedDisruptions := lo.Clamp(minDisruptions-disrupting[nodePool.Name], 0, math.MaxInt32)
```

Comment:
Put also floors for safety reasons, we could just move this floor to zero or max into the dsa itself.